osquery/requests/requests.h

```diff
@@ -0,0 +1,246 @@
+   * set it's own content type. Return a string with the appropriate content
+   * type for serializer.
+   *
+   * @return The content type
+   */
+  virtual std::string getContentType() = 0;
+
+  /**
+   * @brief Serialize a property tree into a string
+   *
+   * @param params A property tree of parameters
+   *
+   * @param serialized The string to populate the final serialized params into
+   *
+   * @return An instance of osquery::Status indicating the success or failure
+   * of the operation
+   */
+  virtual Status serialize(const boost::property_tree::ptree& params,
+                           std::string& serialized) = 0;
+
+  /**
+   * @brief Deerialize a property tree into a property tree
+   *
+   * @param params A string of serialized parameters
+   *
+   * @param serialized The property tree to populate the final serialized
+   * params into
+   *
+   * @return An instance of osquery::Status indicating the success or failure
+   * of the operation
+   */
+  virtual Status deserialize(const std::string& serialized,
+                             boost::property_tree::ptree& params) = 0;
+
+  /**
+   * @brief Virtual destructor
+   */
+  virtual ~Serializer() {}
+
+ protected:
+  /// storage for the transport reference
+  std::shared_ptr<Transport> transport_;
+};
+
+/**
+ * @brief Request class for making flexible remote network requests
+ */
+template <class TTransport, class TSerializer>
+class Request {
+ public:
+  /**
+   * @brief This is the constructor that you should use to instantiate Request
+   *
+   * @param destination A string of the remote URI destination
+   */
+  Request(const std::string& destination) : destination_(destination) {
+    transport_ = std::make_shared<TTransport>();
+    transport_->setDestination(destination_);
+    serializer_ = std::make_shared<TSerializer>();
+
+    transport_->setSerializer(serializer_);
+    serializer_->setTransport(transport_);
+  }
+
+  /**
+   * @brief Class destructor
+   */
+  ~Request() {}
+
+  /**
+   * @brief Send a simple request to the destination with no parameters
+   *
+   * @return An instance of osquery::Status indicating the success or failure
+   * of the operation
+   */
+  Status call() { return transport_->sendRequest(); }
+
+  /**
+   * @brief Send a simple request to the destination with parameters
+   *
+   * @param params A property tree representing the parameters
+   *
+   * @return An instance of osquery::Status indicating the success or failure
+   * of the operation
+   */
+  Status call(const boost::property_tree::ptree& params) {
+    std::string serialized;
+    auto s = serializer_->serialize(params, serialized);
+    if (!s.ok()) {
+      return s;
+    }
+    return transport_->sendRequest(serialized);
+  }
+
+  /**
+   * @brief Get the request response
+   *
+   * @return A pair of a Status and a property tree of response params
+   */
+  std::pair<Status, boost::property_tree::ptree> getResponse() {
```

Comment:
Returning a pair of a status + some output is weird right? How about returning a Status and setting the tree as an output parameter?